tools/decode-packets/src/main/scala/DecodePackets.scala

```diff
@@ -0,0 +1,165 @@
+import java.io.{File, ByteArrayOutputStream, FileOutputStream}
+import java.nio.charset.CodingErrorAction
+import net.psforever.packet.PacketCoding
+import scodec.bits._
+import scodec.Attempt.{Failure, Successful}
+import java.nio.file.{Files, Paths, StandardCopyOption}
+import scala.io.Codec
+import util.control.Breaks._
+import scala.collection.parallel.CollectionConverters._
+import scopt.OParser
+import sys.process._
+
+case class Config(
+    outDir: String = System.getProperty("user.dir"),
+    files: Seq[File] = Seq()
+)
+
+object DecodePackets {
+  def main(args: Array[String]): Unit = {
+
+    val builder = OParser.builder[Config]
+
+    val parser = {
+      import builder._
+      OParser.sequence(
+        programName("psf-decode-packets"),
+        opt[String]('o', "out-dir")
+          .action((x, c) => c.copy(outDir = x))
+          .text("Output directory"),
+        arg[File]("<file>...")
+          .unbounded()
+          .required()
+          .action((x, c) => c.copy(files = c.files :+ x))
+      )
+    }
+
+    val opts = OParser.parse(parser, args, Config()) match {
+      case Some(config) =>
+        config
+      case _ =>
+        sys.exit(1)
+    }
+
+    val outDir = new File(opts.outDir);
+    if(!outDir.exists()) {
+      outDir.mkdirs()
+    } else if(outDir.isFile()) {
+      println(s"error: out-dir is file")
+      sys.exit(1)
+    }
+
+     opts.files.foreach { file =>
+      if(!file.exists) {
+        println(s"file ${file.getAbsolutePath()} does not exist")
+        sys.exit(1)
+      }
+    }
+
+    opts.files.par.foreach { file =>
+      val parsedFile = s"gcapy -xa ${file.getAbsolutePath()}" !!
+      val out = new ByteArrayOutputStream()
```

Comment:
JVM will almost certainly run out of memory if you're processing a folder of all the available gcaps and holding the entire file contents in memory until the end to flush it. Some of the resulting files are in the 300-600MB range.

That was one of the reasons I used the temporary folder, as if anything went wrong during processing for that file you could just clear the temp folder (without having to dig through thousands of log lines to see which files were being processed at the time), then rerun XTools and it wouldn't reprocess already done files when combined with the check to see if the file already exists in the output directory

Reply:
Do you you how many threads it uses? The default thread pool is equal to the number of processors, so if using that, it might be fine?

I wanted to test this by running it over all captures, but I'm getting a lot of errors. Could be related to my Python 3 port of gcapy though, not sure yet.

Reply:
How common are errors anyhow? I didn't really consider them to be normal tbh that's why it's just decoding everything every time.

Reply:
Yeah I believe it's equal to logical processors. Which on my machine is 16 threads. Worst case scenario taking the biggest files all running at the same time that's 2.53GB for the gcapy output, and 5.92GB for the decoded output, plus overheads.

Reply:
Ok so I ran it over all 363 captures and it consumed 3.7GB RAM at most. But I only have 8 logical processors so your estimation is fairly accurate.

So hmm, we could a) limit the number of threads to 4-8 (or make it dynamic: free GB / 500mb) b) have an option to set threads or c) go back to temp files.

I'm not a fan of writing to disk unnecessarily, and you usually don't gain performance by doing that either. But it's probably better than writing to swap..

Reply:
Errors in the actual decoding are fairly rare, but in the event they do happen being able to just exclude that file and not reprocess everything is nice since it does take a while. 

Primarily the errors come from outside factors, for example, the main error that I tend to come across though is actually disk space, on a space constrained SSD it's a bit of a pain to try and clear some space, and have to rerun all over it repeatedly until you have enough space for the whole data set

Reply:
At least with the BufferedWriter you can tweak the length of the buffer it uses, so perhaps increasing that somewhat will alleviate the majority of the performance cost of writing to disk more frequently. Could set the buffer to 16MB or something?

And the temporary files technically aren't temporary really. It's the final output file, just in a temporary location that gets moved so it's not as if the file is being written twice.

Reply:
Also limiting the number of threads seems like a bad solution. The process is primarily CPU bound more than it is disk bound, so that'll slow things down greatly.

Reply:
Yea you're right, revert to BufferedWriter will be up in a sec

Reply:
I also timed the performance and it's maybe 5 or 10 percent that is gained by doing it all in memory. Doesn't matter at all.